From abb2a3c2006f01a39937f80b4c8c60aef81e2e8a Mon Sep 17 00:00:00 2001
From: Scott Wood <scottwood@freescale.com>
Date: Sat, 9 Jun 2012 06:00:40 +0000
Subject: [PATCH 035/227] powerpc/e6500: hw tablewalk: check for NULL pmd
 entry

We need to check for a NULL entry at each level -- the last one was
forgotten.  This resulted in an indirect entry getting loaded pointing
at NULL, which occasionally resulted in CCM errors.

Signed-off-by: Scott Wood <scottwood@freescale.com>
[Kevin: Original patch taken from fsl sdk 1.3.1
QorIQ-SDK-V1.3.1-SOURCE-20121220-yocto.iso.]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 arch/powerpc/mm/tlb_low_64e.S |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/arch/powerpc/mm/tlb_low_64e.S b/arch/powerpc/mm/tlb_low_64e.S
index 1a52e02..bca8d85 100644
--- a/arch/powerpc/mm/tlb_low_64e.S
+++ b/arch/powerpc/mm/tlb_low_64e.S
@@ -338,6 +338,10 @@ tlb_miss_common_fsl_htw:
 	bge	tlb_miss_fault_fsl_htw
 	ldx	r14,r14,r15		/* Grab pmd entry */
 
+	mfspr	r10,SPRN_MAS0
+	cmpdi	cr0,r14,0
+	bge	tlb_miss_fault_fsl_htw
+
 	/* Now we build the MAS for a 2M indirect page:
 	 *
 	 * MAS 0   :	ESEL needs to be filled by software round-robin
@@ -349,7 +353,6 @@ tlb_miss_common_fsl_htw:
 	 */
 
 
-	mfspr	r10,SPRN_MAS0
 	rldicr	r16,r11,0,62
 	lwz	r15,0(r16)
 
-- 
1.7.9.7

